### PR TITLE
List companies by category.

### PR DESCRIPTION
**Pull request by thesuss**
PT Story: https://www.pivotaltracker.com/story/show/135684057

Changes proposed in this pull request:
1. business_category_path lists containing companies
2. Routes are in swedish (see another PR)

Screenshots (Optional):
skarmavbild 2016-12-16 kl 23 44 45

WIP: because: Too much logic for a view. Please help. 
Also if there is a way to show a message when no companies found that would be helpful. 
(didn't figure out how to do that one)

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -10,11 +10,13 @@
     end
   end
 
-  get "/pages/*id" => 'pages#show', as: :page, format: false
+  get '/pages/*id', to: 'pages#show', as: :page, format: false
 
-  resources :business_categories
-  resources :membership_applications, only: [:new, :create, :edit, :update, :index, :show]
-  resources :companies, only: [:new, :create, :edit, :update, :index, :show]
+  scope(path_names: { new: 'ny', edit: 'redigera' }) do
```
Comment (weedySeaDragon) [inline config/routes.rb]: This should be a separate PR.  It could be part of the I18n work if you want to refer to that PT entry.

File: features/list_companies_of_category.feature [hunk 2]
```diff
@@ -0,0 +1,36 @@
+Feature: As a visitor
+  In order to easily find a member of a certain category
```
Comment (weedySeaDragon) [inline features/list_companies_of_category.feature]: Use the word 'company' instead of 'member'; "member" = a person that is a member of SHF.

**Comment by weedySeaDragon** [comment]
Let the Rails & the database do the work:  this is what it's good at.  (looking up data and the connections).

Re: the feature -- You should always try to check 0, 1, and more than one of something.
- What is the expected behavior if there are no companies for a category?  (should this happen?  can it happen if a member with category "blorf" existed, but then the member was deleted?)
- add in another member/company so that you have a scenario that shows more than 1 company in the list for a category
- what happens if the categories for a member are changed?  are the right companies shown in the right places? (the companies _should_ show up on the changed categories)

File: app/views/business_categories/show.html.haml [hunk 3]
```diff
@@ -1,15 +1,32 @@
...
+        - if company_complete? company
+          - company.categories.each do |category|
+            - if category.name == @business_category.name
+              %tr.company
+                %td= link_to company.name, company
+                %td= company.region
+                - if current_user.try(:admin)
+                  %td= company.company_number
+                  %td= link_to 'Edit', edit_company_path(company)
+                  %td= link_to 'Delete', company, method: :delete, data: { :confirm => 'Are you sure?' }
```
Comment (patmbolger) [inline app/views/business_categories/show.html.haml]: I have trouble following this logic.  

The Company model has this:
```
has_and_belongs_to_many :business_categories
```
but the BusinessCategory does not have the same association back to Company.

Here, we are looping through the categories associated with the employees of the company (via `Company#categories`).

So, it is unclear to me whether a Company should have a direct association with BusinessCategory, or no association and only be indirectly linked via employee's business categories, or both(?).

In the DB I am using, "Happy Mutts" has zero categories via employees and 2 categories directly:
```
2.3.1 :061 >   hm = Company.find_by_name('Happy Mutts')
 => #<Company id: 3, name: "Happy Mutts", company_number: "5562252998", phone_number: "", email: "kicki@gladajyckar.se", street: "Ålstensgatan 4", post_code: "123 45", city: "Bromma", region: "Stockholm", website: "http://www.gladajyckar.se", created_at: "2016-12-01 14:27:05", updated_at: "2016-12-01 14:27:05"> 
2.3.1 :062 > 
2.3.1 :063 >   hm.categories.count
 => 0 
2.3.1 :064 > hm.business_categories.count
 => 2 
2.3.1 :065 > 
```
Happy Mutts, of course, shows zero categories in this `show` view.

Do we need/want two types of business category associations for a company?  Please advise.

**Review by patmbolger — COMMENTED** [review]
Please see comment/question.

File: app/views/business_categories/show.html.haml (hunk 3 above)
Comment (thesuss) [reply app/views/business_categories/show.html.haml]: I can't answer in your language, but the company gets its categories via its members. So if I'm approved to be a groomer and trainer, those should be listed for my company, via my application. 

If I employ someone who is a psycologist, I cant just choose that category, but they need to apply for membership in that category, for my company.

Comment (thesuss) [reply app/views/business_categories/show.html.haml]: Try dropping and recreating your db (if you're in rails c), it could be an old leftover.

Comment (thesuss) [reply app/views/business_categories/show.html.haml]: Maybe it's related to: https://www.pivotaltracker.com/n/projects/1904891/stories/136300427 ?

Comment (weedySeaDragon) [reply app/views/business_categories/show.html.haml]: That association needs to be removed.  The association between a company and business categories is a remnant from the 'first' descriptions of the system, when _companies_ were the entities that were applying for membership.  That changed and now it's _individuals_ [users] that apply for membership (with a membership_application).


The associations are:
-  a user has (fills out) a membership application
   -  a membership application has many business categories (these are the skills/business services that a user is certified to offer)
        - a membership application has a company; the user is employed by the company

- a company can have many membership_applications (the mirror of the above)


I poked around at this, and this is what I've come up with:
1.  In`BusinessCategory` we need to add `has_many :membership_applications` so we can use the association from that direction
2. In `Company` we need to add `has_many :membership_applications`
2. in `BusinessCategoriesController`:
``` 
 def show
    @companies = Company.includes(membership_applications: :business_categories).where(business_categories: { id: @business_category.id }).references(:business_categories)
  end
```

In the view, we can then loop through @companies and display each one.

I was going to talk with @thesuss about this before doing anything with my code. (I thought she might want to do it after we went over it.)  But since you asked, Pat, I thought I'd go ahead and respond with it here. :-)

_(BTW -- the line in the above `show` method is the kind of thing I should be using in `Company#categories`.  ref: #82 )_

Comment (patmbolger) [reply app/views/business_categories/show.html.haml]: Ok, that makes sense.  I suggest getting the model associations straightened out ASAP to prevent confusion and coding errors. (or, maybe just my confusion :) )

Comment (patmbolger) [reply app/views/business_categories/show.html.haml]: @weedySeaDragon - should ```BusinessCategory``` also ```have_many :companies, through: :membership_applications```?  

That would make the `show` view in this PR much simpler, since we are starting from category and want to list the companies (we'll have to filter for duplicates, I think, since I suppose it's possible for two or more member_applications to reference the same company(?)).

Comment (weedySeaDragon) [reply app/views/business_categories/show.html.haml]: yes -- that makes good sense.

I think I should start another PR and  -- once I've thought through these more -- will do these changes to the models so that we have the associations cleaned up.  Then this branch can be merged and will be easier.

I'll likely do it tomorrowish (PST).

File: features/list_companies_of_category.feature [hunk 4]
```diff
@@ -0,0 +1,36 @@
+Feature: As a visitor
```
Comment (weedySeaDragon) [inline features/list_companies_of_category.feature]: You must be explicit about whether someone is logged in or not for a scenario.  Always have either 'I am logged out' or 'I am logged in as "...."'

Please add scenarios so we have all of theses possibilities covered:

- A visitor views a category
- A user views a category
- A member views a category
- An admin views a category

- A visitor tries to view a category with no companies
- A visitor views a category with 3 companies (You should verify that they see _3_ companies)



Remember that this _feature_ should describe what happens in all cases, for all types of users.   We need a description of what happens in all possibilities so that:
- we know we've thought through the feature completely,
- others reading this in the future and trying  to understand what should happen with this feature get a complete picture,
**and importantly:**
- to be _sure_ that in the future as the system grows in complexity, we are testing everything.

I know that the Feature description starts off with "As a visitor..."  but that doesn't mean we only test that narrow possibility.  We must always test a full range of possibilities.

File: app/views/business_categories/show.html.haml [hunk 5]
```diff
@@ -1,15 +1,28 @@
...
+  %table
+    %thead
+      %tr
+        %th Namn
+        %th Län
+        - if current_user.try(:admin)
+          %th Org nr
+          %th
+    %tbody
+      - @business_category.companies.each do |company|
```
Comment (patmbolger) [inline app/views/business_categories/show.html.haml]: You may want to get unique companies within category (because we are going through membership_applications to get companies, and one company could occur in more than one membership_application).

Also, may want to sort the companies by name (not a big deal until we have a lot of companies, of course).

Something like:
```
@business_category.companies.distinct.sort {|cp1, cp2| cp1.name <=> cp2.name}.each do |company|
```

**Review by weedySeaDragon — CHANGES_REQUESTED** [review]

**Review by patmbolger — COMMENTED** [review]
Looks good ... please see comment.

File: features/list_companies_of_category.feature (hunk 4 above)
Comment (thesuss) [reply features/list_companies_of_category.feature]: I don't see why we need to test if it is visible to **everyone**, in the agile way that is beyond the scope of this feature. If admin doesn't see it, that is irrelevant for this feature. 
Then the Feature needs to be better described (As `Anyone` or as `Visitor, member and admin`). Right?

I do see why we probably should test 3

Comment (weedySeaDragon) [reply features/list_companies_of_category.feature]: Yes -- you're seeing it. :-)  And there's a difference between describing a feature and doing the testing.  

The crux of it is: the feature needs to say "everyone" -- and so you also need to test for everyone.  (If the feature meant only visitors, then you have to test to ensure that only visitors can do it, and that users, members, and admins cannot.)

File: app/views/business_categories/show.html.haml [hunk 6]
```diff
@@ -1,15 +1,28 @@
...
+  %table
+    %thead
+      %tr
+        %th Namn
+        %th Län
+        - if current_user.try(:admin)
+          %th Org nr
+          %th
+    %tbody
+      - @business_category.companies.distinct.sort {|cp1, cp2| cp1.name <=> cp2.name}.each do |company|
```
Comment (patmbolger) [inline app/views/business_categories/show.html.haml]: A simple refactor story would be to move this logic (```.companies.distinct.sort {|cp1, cp2| cp1.name <=> cp2.name}``` into an instance method in the BusinessCategory model.

File: features/list_companies_of_category.feature (hunk 4 above)
Comment (thesuss) [reply features/list_companies_of_category.feature]: added for admin and for logged in user

**Review by weedySeaDragon — APPROVED** [review]
